From: Andi Kleen <ak@linux.intel.com>
Subject: HWMON: Convert coretemp to x86 cpuid autoprobing
References: http://lists.opensuse.org/opensuse-kernel/2011-11/msg00075.html
Patch-Mainline: 9b38096fde5f9b93c3657911c3be7892cc155cbd undefined


Signed-off-by: Thomas Renninger <trenn@suse.de>

Use the new x86 cpuid autoprobe interface for the Intel coretemp
driver.

Cc: Fenghua Yu <fenghua.yu@intel.com>
Cc: Jean Delvare <khali@linux-fr.org>
Cc: Guenter Roeck <guenter.roeck@ericsson.com>
Signed-off-by: Andi Kleen <ak@linux.intel.com>
Signed-off-by: Thomas Renninger <trenn@suse.de>
Acked-by: H. Peter Anvin <hpa@zytor.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

Automatically created from "patches.arch/006-x86_autoload.patch" by xen-port-patches.py

--- head.orig/drivers/hwmon/coretemp-xen.c	2011-11-17 16:53:49.000000000 +0100
+++ head/drivers/hwmon/coretemp-xen.c	2012-02-16 14:36:30.000000000 +0100
@@ -38,6 +38,7 @@
 #include <linux/smp.h>
 #include <linux/moduleparam.h>
 #include <asm/msr.h>
+#include <asm/cpu_device_id.h>
 #include <xen/pcpu.h>
 #include "../xen/core/domctl.h"
 
@@ -819,6 +820,12 @@ static struct notifier_block coretemp_cp
 	.notifier_call = coretemp_cpu_callback,
 };
 
+static const struct x86_cpu_id coretemp_ids[] = {
+	{ X86_VENDOR_INTEL, X86_FAMILY_ANY, X86_MODEL_ANY, X86_FEATURE_DTS },
+	{}
+};
+MODULE_DEVICE_TABLE(x86cpu, coretemp_ids);
+
 static int __init coretemp_init(void)
 {
 	int err = -ENODEV;
@@ -826,9 +833,13 @@ static int __init coretemp_init(void)
 	if (!is_initial_xendomain())
 		goto exit;
 
-	/* quick check if we run Intel */
-	if (cpu_data(0).x86_vendor != X86_VENDOR_INTEL)
-		goto exit;
+	/*
+	 * CPUID.06H.EAX[0] indicates whether the CPU has thermal
+	 * sensors. We check this bit only, all the early CPUs
+	 * without thermal sensors will be filtered out.
+	 */
+	if (!x86_match_cpu(coretemp_ids))
+		return -ENODEV;
 
 	err = platform_driver_register(&coretemp_driver);
 	if (err)
